File: go/vulnerable.go
/**
 * Vulnerable Go code samples for SAST testing
 * Author: Test Suite
 * Created: 2025-01-21
 */

package main

import (
	"crypto/md5"
	"database/sql"
	"fmt"
	"io/ioutil"
	"os"
	"os/exec"
	"path/filepath"
)

// SQL Injection vulnerability
func getUserData(userID string) (*sql.Rows, error) {
	query := fmt.Sprintf("SELECT * FROM users WHERE id = %s", userID)
	return db.Query(query)
}

// Command Injection
func executeCommand(command string) error {
	cmd := exec.Command("sh", "-c", command)
	return cmd.Run()
}

// Hardcoded secrets
const (
	apiKey    = "sk_live_51H3ll0W0rld123456789"
	password  = "admin123"
	secretKey = "mySecretKey123"
)

// Path Traversal
func readFile(filename string) ([]byte, error) {
	path := filepath.Join("/var/www/data", filename)
	return ioutil.ReadFile(path)
}

// Weak crypto - MD5
func hashPassword(password string) string {
	hash := md5.Sum([]byte(password))
	return fmt.Sprintf("%x", hash)
}

// Insecure random
func generateToken() string {
	return fmt.Sprintf("%d", os.Getpid())
}

// Command Injection with exec.Command
func runCommand(cmd string) error {
	return exec.Command("bash", "-c", cmd).Run()
}

File: swift/vulnerable.swift
/**
 * Vulnerable Swift code samples for SAST testing
 * Author: Test Suite
 * Created: 2025-01-21
 */

import Foundation
import CommonCrypto

// SQL Injection vulnerability (if using string concatenation)
func getUserData(userId: String) -> String {
    return "SELECT * FROM users WHERE id = \(userId)"
}

// Command Injection
func executeCommand(_ command: String) {
    let process = Process()
    process.launchPath = "/bin/sh"
    process.arguments = ["-c", command]
    process.launch()
}

// Hardcoded secrets
let API_KEY = "sk_live_51H3ll0W0rld123456789"
let PASSWORD = "admin123"
let SECRET = "mySecretKey123"

// Path Traversal
func readFile(_ filename: String) throws -> String {
    let path = "/var/www/data/\(filename)"
    return try String(contentsOfFile: path, encoding: .utf8)
}

// Weak crypto - MD5
func hashPassword(_ password: String) -> String {
    let data = password.data(using: .utf8)!
    var digest = [UInt8](repeating: 0, count: Int(CC_MD5_DIGEST_LENGTH))
    data.withUnsafeBytes {
        _ = CC_MD5($0.baseAddress, CC_LONG(data.count), &digest)
    }
    return digest.map { String(format: "%02x", $0) }.joined()
}

// Insecure random
func generateToken() -> String {
    return String(arc4random())
}

// Command Injection with Process
func runCommand(_ cmd: String) {
    let task = Process()
    task.launchPath = "/usr/bin/env"
    task.arguments = ["bash", "-c", cmd]
    task.launch()
}

